client/chat/channel-list.tsx

```diff
@@ -0,0 +1,168 @@
+import { useAutoFocusRef } from '../material/auto-focus'
+import { RaisedButton } from '../material/button'
+import { useLocationSearchParam } from '../navigation/router-hooks'
+import { push } from '../navigation/routing'
+import { useRefreshToken } from '../network/refresh-token'
+import { useAppDispatch } from '../redux-hooks'
+import { SearchInput, SearchInputHandle } from '../search/search-input'
+import { useStableCallback } from '../state-hooks'
+import { colorError, colorTextFaint } from '../styles/colors'
+import { headline4, subtitle1 } from '../styles/typography'
+import { searchChannels } from './action-creators'
+
+const SEARCH_CHANNELS_LIMIT = 40
+
+const Container = styled.div`
+  padding: 16px 24px;
+
+  display: flex;
+  flex-direction: column;
+`
+
+const TitleBar = styled.div`
+  display: flex;
+  flex-direction: row;
+  justify-content: space-between;
+  align-items: center;
+  margin-bottom: 16px;
+`
+
+const PageHeadline = styled.div`
+  ${headline4};
+`
+
+const StyledSearchInput = styled(SearchInput)`
+  width: 256px;
+`
+
+const SearchResults = styled.div`
+  display: flex;
+  flex-wrap: wrap;
+  gap: 8px;
+  margin: 16px 0;
+`
+
+const NoResults = styled.div`
+  ${subtitle1};
+
+  color: ${colorTextFaint};
+`
+
+const ErrorText = styled.div`
+  ${subtitle1};
+
+  color: ${colorError};
+`
+
+export function ChannelList() {
+  const dispatch = useAppDispatch()
+  const autoFocusRef = useAutoFocusRef<SearchInputHandle>()
+
+  const [channels, setChannels] = useState<BasicChannelInfo[]>()
+  const [hasMoreChannels, setHasMoreChannels] = useState(true)
+
+  const [isLoadingMoreChannels, setIsLoadingMoreChannels] = useState(false)
+  const [searchError, setSearchError] = useState<Error>()
+  const [searchQuery, setSearchQuery] = useLocationSearchParam('q')
+  const abortControllerRef = useRef<AbortController>()
+
+  const [refreshToken, triggerRefresh] = useRefreshToken()
+  const debouncedSearchRef = useRef(
+    debounce((searchQuery: string) => {
+      // Just need to clear the search results here and let the carousel initiate the network
+      // request.
+      setSearchQuery(searchQuery)
+      setSearchError(undefined)
+      setChannels(undefined)
+      setHasMoreChannels(true)
+      triggerRefresh()
+    }, 100),
+  )
+
+  const onSearchChange = useStableCallback((searchQuery: string) => {
+    debouncedSearchRef.current(searchQuery)
+  })
+
+  const onLoadMoreChannels = useStableCallback(() => {
+    setIsLoadingMoreChannels(true)
+
+    abortControllerRef.current?.abort()
+    abortControllerRef.current = new AbortController()
+
+    dispatch(
+      searchChannels(searchQuery, SEARCH_CHANNELS_LIMIT, channels?.length ?? 0, {
+        signal: abortControllerRef.current.signal,
+        onSuccess: data => {
+          setIsLoadingMoreChannels(false)
+          setChannels((channels ?? []).concat(data.channelInfos))
+          setHasMoreChannels(data.channelInfos.length >= SEARCH_CHANNELS_LIMIT)
```

Comment:
Like I said before, it would be better if any paginated APIs just returned this information in their response instead of doing stuff like this. The service can implement this in a similar way, it just makes it so we don't have to keep and re-use this value on the client to do this check (and the server may have a better idea of how many results there actually are, so it can avoid the client requesting empty lists in some cases)

I understand that some of these APIs existed before, but the search one is new, so I don't know why we're repeating bad design choices in it (and in some cases coming up with brand new bad design choices)